include/msgpack/v1/adaptor/string.hpp

```diff
@@ -30,10 +30,16 @@ struct convert<std::string> {
     msgpack::object const& operator()(msgpack::object const& o, std::string& v) const {
         switch (o.type) {
         case msgpack::type::BIN:
-            v.assign(o.via.bin.ptr, o.via.bin.size);
+            if (o.via.bin.size == 0)
+                v.clear();
+            else
+                v.assign(o.via.bin.ptr, o.via.bin.size);
```

Comment:
This is similar code to `memcmp()` we discussed.
But I think it is not requred.

In this case, both valgrind and sanitizer don't report error.

Reply:
libc++ implementation is 
https://github.com/llvm/llvm-project/blob/main/libcxx/include/string#L1909-L1919

libstdc++ implementation is
https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/bits/basic_string.h#L1440-L1445
https://github.com/gcc-mirror/gcc/blob/16e2427f50c208dfe07d07f18009969502c25dc8/libstdc%2B%2B-v3/include/debug/debug.h#L122-L123
https://github.com/gcc-mirror/gcc/blob/16e2427f50c208dfe07d07f18009969502c25dc8/libstdc%2B%2B-v3/include/debug/macros.h#L462-L463

Reply:
I will update the msgpack code and check it using cmake -T memcheck soon.

Reply:
I've tested both sanitizer undefined checking and valgrind for all tests on my local environment, and got no error.

Reply:
I tried to revert the change, and now CI build fails: https://github.com/kovdan01/msgpack-c/pull/9/checks?check_run_id=3450115223.

It is also reproduced on my local environment (`ctest -T memcheck -R user_class`). More verbose output:

```
==16998== Memcheck, a memory error detector
==16998== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==16998== Using Valgrind-3.17.0 and LibVEX; rerun with -h for copyright info
==16998== Command: /home/kovdan01/Projects/msgpack-c/build-problem/test/user_class
==16998== Parent PID: 16997
==16998==
==16998== Conditional jump or move depends on uninitialised value(s)
==16998==    at 0x4AB008B: _M_disjunct (basic_string.h:340)
==16998==    by 0x4AB008B: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_replace(unsigned long, unsigned long, char const*, unsigned long) (basic_string.tcc:462)
==16998==    by 0x14D212: msgpack::v1::adaptor::convert<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, void>::operator()(msgpack::v1::object const&, std::__cxx11::basic_stri>
==16998==    by 0x15B387: msgpack::v1::enable_if<!msgpack::v1::is_array<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::value, msgpack::v1::object const&>::type msgpack::v1>
==16998==    by 0x158AD0: msgpack::v1::enable_if<(!msgpack::v1::is_array<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::value)&&(!msgpack::v1::is_pointer<std::__cxx11::bas>
==16998==    by 0x155BAB: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > msgpack::v1::object::as<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>
==16998==    by 0x14EA1C: msgpack::v1::adaptor::convert<test_non_intrusive, void>::operator()(msgpack::v1::object const&, test_non_intrusive&) const (user_class.cpp:503)
==16998==    by 0x15CC4D: msgpack::v1::enable_if<!msgpack::v1::is_array<test_non_intrusive>::value, msgpack::v1::object const&>::type msgpack::v1::operator>><test_non_intrusive>(msgpack::v1::object const&, t>
==16998==    by 0x15A4EE: msgpack::v1::enable_if<(!msgpack::v1::is_array<test_non_intrusive>::value)&&(!msgpack::v1::is_pointer<test_non_intrusive>::value), test_non_intrusive&>::type msgpack::v1::object::co>
==16998==    by 0x155CE7: test_non_intrusive msgpack::v1::object::as<test_non_intrusive>() const (object.hpp:1114)
==16998==    by 0x145979: test_non_intrusive_::test_method() (user_class.cpp:542)
==16998==    by 0x1455A3: test_non_intrusive__invoker() (user_class.cpp:534)
==16998==    by 0x15D439: boost::detail::function::void_function_invoker0<void (*)(), void>::invoke(boost::detail::function::function_buffer&) (function_template.hpp:117)
==16998==
==16998==
==16998== HEAP SUMMARY:
==16998==     in use at exit: 0 bytes in 0 blocks
==16998==   total heap usage: 31,446 allocs, 31,446 frees, 58,095,399 bytes allocated
==16998==
==16998== All heap blocks were freed -- no leaks are possible
==16998==
==16998== Use --track-origins=yes to see where uninitialised values come from
==16998== For lists of detected and suppressed errors, rerun with: -s
==16998== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)

```

I am unable to detect the problem with sanitizers, looks like it is only detected with valgrind.

Reply:
It is very weird. I don't understand that why that special case is failed.
I want to know why it happens.
How about minimal example like https://github.com/msgpack/msgpack-c/pull/951#discussion_r697843338 with the same compiler setting and the same valgrind setting?
I will also check it when I have a time.

Reply:
OK, I'll try to create a minimal example that represents the problem. I'll inform you about the results.

Reply:
It is not minimal enough but I found one reproduce code:

```cpp
#include <cassert>
#include <string>

struct object;

struct object_str {
    uint32_t size;
    const char* ptr;
};

struct object_bin {
    uint32_t size;
    const char* ptr;
};

struct object_array {
    uint32_t size;
    object* ptr;
};

struct object {
    union {
        object_str str;
        object_bin bin;
        object_array array;
    } via;
};

inline void convert(object const& o, std::string& v) {
    v.assign(o.via.str.ptr, o.via.str.size);
}

std::string  as(object const& o) {
    std::string v;
    convert(o, v);
    return v;
}

struct my {
    void set_name(std::string const& name) {
        this->name = name;
    }
    std::string name;
};

int main() {
    char const* p1 = (char*)(0x1);
    object o;
    o.via.str.size = 0;
    // o.via.str.ptr = nullptr;
    my m;
    m.set_name(as(o));
}
```

If I uncomment the null initialize, then the report is disappeared.

Reply:
The reporting happens only if MSGPACK_API_VERSION=1. It is a important point.

### API_VERSION 1 unpack

https://github.com/msgpack/msgpack-c/blob/be4d971c62798eb59f8455dc77a4529748bcd08f/include/msgpack/v1/unpack.hpp#L167-L197

### API_VERION >= 2 unpack

https://github.com/msgpack/msgpack-c/blob/be4d971c62798eb59f8455dc77a4529748bcd08f/include/msgpack/v2/create_object_visitor.hpp#L110-L157

### Hypothesis
When string size is 0, on API_VERSION 1 unpack's `o.via.str.ptr` is uninitialized but API_VERSION >=2 unpack's `o.via.str.ptr` is initialized as nullptr.

I think that it is essential reason.

Reply:
Could you test the following fix?
If it works, please add the fix to this PR. And remove size check from assign().

```diff
diff --git a/include/msgpack/v1/unpack.hpp b/include/msgpack/v1/unpack.hpp
index 4f5dab47..fefee66d 100644
--- a/include/msgpack/v1/unpack.hpp
+++ b/include/msgpack/v1/unpack.hpp
@@ -177,6 +177,9 @@ inline void unpack_str(unpack_user& u, const char* p, uint32_t l, msgpack::objec
         std::memcpy(tmp, p, l);
         o.via.str.ptr = tmp;
     }
+    else {
+        o.via.str.ptr = MSGPACK_NULLPTR;
+    }
     o.via.str.size = l;
 }

@@ -193,6 +196,9 @@ inline void unpack_bin(unpack_user& u, const char* p, uint32_t l, msgpack::objec
         std::memcpy(tmp, p, l);
         o.via.bin.ptr = tmp;
     }
+    else {
+        o.via.bin.ptr = MSGPACK_NULLPTR;
+    }
     o.via.bin.size = l;
 }
```

Reply:
The fix works (both locally and on CI), thanks! You are right – the root cause was in uninitialized ptr value. Updated the PR (added the appropriate fix & deleted unneeded size check in assign code).